### PR TITLE
Resolves #1124: Always scan the repository for provided scripts

### LINKED ISSUE #1124: s2i build --as-dockerfile not using local .s2i/[assemble, run] when builder image present
When running s2i build with both local script files and a valid builder image, I'm not seeing the resulting
dockerfile having the correct RUN and CMD statements. However, when pointing to an invalid builder image
the expected set of statements and other elements (in the debug log) appear present in the container build file.

**Version:** 1.3.2 (Fedora 38) and 1.3.8

**Reproduction:**

(This repository contains both .s2i/bin/run and .s2i/bin/assemble

```
$ git clone https://github.com/cuppett/laravel-s2i.git
$ cd laravel-s2i
$ s2i build . registry.access.redhat.com/ubi8/php-80 --as-dockerfile Containerfile.invalid --loglevel=5
```

```log
I0807 07:26:59.242962  100612 build.go:52] Running S2I version "v1.3.8"
I0807 07:26:59.243069  100612 util.go:70] Getting docker credentials for registry.access.redhat.com/ubi8/php-80
I0807 07:27:00.156576  100612 build.go:170] 
Source:				.
Output Image Tag:		
Environment:			
Labels:				
Incremental Build:		disabled
Remove Old Build:		disabled
Builder Pull Policy:		if-not-present
Previous Image Pull Policy:	if-not-present
Quiet:				disabled
Layered Build:			disabled
Docker Endpoint:		unix:///run/user/1000/podman/podman.sock
Docker Pull Config:		/home/scuppett/.docker/config.json
Docker Pull User:		

I0807 07:27:00.156769  100612 scm.go:20] DownloadForSource .
I0807 07:27:00.157914  100612 clone.go:37] Downloading "." ...
I0807 07:27:00.157922  100612 clone.go:41] Cloning sources into "upload/src"
I0807 07:27:00.198650  100612 clone.go:57] Checked out "HEAD"
I0807 07:27:00.213746  100612 clone.go:63] Updated submodules for "HEAD"
I0807 07:27:00.219627  100612 install.go:269] Using "assemble" installed from "<source-dir>/.s2i/bin/assemble"
I0807 07:27:00.219660  100612 install.go:269] Using "run" installed from "<source-dir>/.s2i/bin/run"
I0807 07:27:00.219680  100612 install.go:269] Using "save-artifacts" installed from "image:///usr/libexec/s2i/save-artifacts"
I0807 07:27:00.219698  100612 ignore.go:84] .s2iignore lists a file spec of .dockerignore 
I0807 07:27:00.219703  100612 ignore.go:116] using globspec upload/src/.dockerignore 
I0807 07:27:00.219711  100612 ignore.go:122] Glob result upload/src/.dockerignore 
I0807 07:27:00.219715  100612 ignore.go:84] .s2iignore lists a file spec of .env* 
I0807 07:27:00.219720  100612 ignore.go:116] using globspec upload/src/.env* 
I0807 07:27:00.219751  100612 ignore.go:122] Glob result upload/src/.env.example 
I0807 07:27:00.219756  100612 ignore.go:84] .s2iignore lists a file spec of .git 
I0807 07:27:00.219761  100612 ignore.go:116] using globspec upload/src/.git 
I0807 07:27:00.219768  100612 ignore.go:122] Glob result upload/src/.git 
I0807 07:27:00.219772  100612 ignore.go:84] .s2iignore lists a file spec of .idea 
I0807 07:27:00.219776  100612 ignore.go:116] using globspec upload/src/.idea 
I0807 07:27:00.219785  100612 ignore.go:84] .s2iignore lists a file spec of .s2iignore 
I0807 07:27:00.219789  100612 ignore.go:116] using globspec upload/src/.s2iignore 
I0807 07:27:00.219796  100612 ignore.go:122] Glob result upload/src/.s2iignore 
I0807 07:27:00.219800  100612 ignore.go:84] .s2iignore lists a file spec of Dockerfile.s2i-fake 
I0807 07:27:00.219804  100612 ignore.go:116] using globspec upload/src/Dockerfile.s2i-fake 
I0807 07:27:00.219811  100612 ignore.go:84] .s2iignore lists a file spec of LICENSE 
I0807 07:27:00.219816  100612 ignore.go:116] using globspec upload/src/LICENSE 
I0807 07:27:00.219823  100612 ignore.go:84] .s2iignore lists a file spec of README.md 
I0807 07:27:00.219827  100612 ignore.go:116] using globspec upload/src/README.md 
I0807 07:27:00.219834  100612 ignore.go:122] Glob result upload/src/README.md 
I0807 07:27:00.219839  100612 ignore.go:84] .s2iignore lists a file spec of vendor 
I0807 07:27:00.219843  100612 ignore.go:116] using globspec upload/src/vendor 
I0807 07:27:00.219853  100612 ignore.go:45] attempting to remove file upload/src/.dockerignore 
I0807 07:27:00.219883  100612 ignore.go:45] attempting to remove file upload/src/.env.example 
I0807 07:27:00.219905  100612 ignore.go:45] attempting to remove file upload/src/.git 
I0807 07:27:00.223812  100612 ignore.go:45] attempting to remove file upload/src/.s2iignore 
I0807 07:27:00.223836  100612 ignore.go:45] attempting to remove file upload/src/README.md 
I0807 07:27:00.223855  100612 dockerfile.go:115] Constructing image build context directory at .
I0807 07:27:00.223893  100612 environment.go:41] Setting 7 environment variables provided by environment file in sources
I0807 07:27:00.223912  100612 dockerfile.go:223] Processing injected inputs: api.VolumeList(nil)
I0807 07:27:00.223918  100612 dockerfile.go:225] Processed injected inputs: api.VolumeList(nil)
I0807 07:27:00.223969  100612 dockerfile.go:296] Wrote custom Dockerfile to Containerfile.invalid
I0807 07:27:00.223977  100612 build.go:180] Application dockerfile generated in Containerfile.invalid
```

`$ cat Containerfile.invalid `

```dockerfile
FROM registry.access.redhat.com/ubi8/php-80:latest
LABEL "io.openshift.s2i.build.commit.date"="Sun Mar 20 08:16:49 2022 -0400" \
      "io.openshift.s2i.build.commit.id"="fefad4475746a38a4197f36bff312dd056155f48" \
      "io.openshift.s2i.build.commit.ref"="master" \
      "io.openshift.s2i.build.commit.message"="Upgrade Laravel 9 (Laravel 8.68.1 -> 9.5.1)" \
      "io.openshift.s2i.build.source-location"="/home/scuppett/Development/php/laravel-s2i/." \
      "io.openshift.s2i.build.image"="registry.access.redhat.com/ubi8/php-80:latest" \
      "io.openshift.s2i.build.commit.author"="Stephen Cuppett <steve@cuppett.com>"
ENV DOCUMENTROOT="/public" \
    APP_NAME="Laravel" \
    APP_DEBUG="true" \
    APP_URL="http://localhost:8080" \
    LOG_CHANNEL="errorlog" \
    DB_CONNECTION="sqlite" \
    DB_DATABASE=":memory:"
USER root
# Copying in source code
COPY upload/src /tmp/src
# Change file ownership to the assemble user. Builder image must support chown command.
RUN chown -R 1001:0 /tmp/src
USER 1001
# Assemble script sourced from builder image based on user input or image metadata.
# If this file does not exist in the image, the build will fail.
RUN /usr/libexec/s2i/assemble
# Run script sourced from builder image based on user input or image metadata.
# If this file does not exist in the image, the build will fail.
CMD /usr/libexec/s2i/run
```

I would expect the RUN and CMD to point to the local files, not the originals for the builder image.

Using a nonexistant builder image (with the same repository) yields what I believe is the correct results.

`$ s2i build . registry.access.redhat.com/ubi8/php-83 --as-dockerfile Containerfile.valid --loglevel=5`

```log
I0807 07:29:37.883152  100867 build.go:52] Running S2I version "v1.3.8"
I0807 07:29:37.883197  100867 util.go:70] Getting docker credentials for registry.access.redhat.com/ubi8/php-83
W0807 07:29:38.222461  100867 build.go:167] could not inspect the builder image for labels: reading manifest latest in registry.access.redhat.com/ubi8/php-83: name unknown: Repo not found
I0807 07:29:38.222513  100867 build.go:170] 
Source:				.
Output Image Tag:		
Environment:			
Labels:				
Incremental Build:		disabled
Remove Old Build:		disabled
Builder Pull Policy:		if-not-present
Previous Image Pull Policy:	if-not-present
Quiet:				disabled
Layered Build:			disabled
Docker Endpoint:		unix:///run/user/1000/podman/podman.sock
Docker Pull Config:		/home/scuppett/.docker/config.json
Docker Pull User:		

I0807 07:29:38.222701  100867 install.go:269] Using "assemble" installed from "image:///usr/libexec/s2i/assemble"
I0807 07:29:38.222715  100867 install.go:269] Using "run" installed from "image:///usr/libexec/s2i/run"
I0807 07:29:38.222724  100867 install.go:269] Using "save-artifacts" installed from "image:///usr/libexec/s2i/save-artifacts"
I0807 07:29:38.222732  100867 scm.go:20] DownloadForSource .
I0807 07:29:38.223926  100867 clone.go:37] Downloading "." ...
I0807 07:29:38.223934  100867 clone.go:41] Cloning sources into "upload/src"
I0807 07:29:38.240659  100867 clone.go:57] Checked out "HEAD"
I0807 07:29:38.255107  100867 clone.go:63] Updated submodules for "HEAD"
I0807 07:29:38.261764  100867 install.go:269] Using "assemble" installed from "<source-dir>/.s2i/bin/assemble"
I0807 07:29:38.261790  100867 install.go:269] Using "run" installed from "<source-dir>/.s2i/bin/run"
I0807 07:29:38.261810  100867 ignore.go:84] .s2iignore lists a file spec of .dockerignore 
I0807 07:29:38.261815  100867 ignore.go:116] using globspec upload/src/.dockerignore 
I0807 07:29:38.261821  100867 ignore.go:122] Glob result upload/src/.dockerignore 
I0807 07:29:38.261825  100867 ignore.go:84] .s2iignore lists a file spec of .env* 
I0807 07:29:38.261829  100867 ignore.go:116] using globspec upload/src/.env* 
I0807 07:29:38.261853  100867 ignore.go:122] Glob result upload/src/.env.example 
I0807 07:29:38.261857  100867 ignore.go:84] .s2iignore lists a file spec of .git 
I0807 07:29:38.261860  100867 ignore.go:116] using globspec upload/src/.git 
I0807 07:29:38.261865  100867 ignore.go:122] Glob result upload/src/.git 
I0807 07:29:38.261868  100867 ignore.go:84] .s2iignore lists a file spec of .idea 
I0807 07:29:38.261871  100867 ignore.go:116] using globspec upload/src/.idea 
I0807 07:29:38.261877  100867 ignore.go:84] .s2iignore lists a file spec of .s2iignore 
I0807 07:29:38.261880  100867 ignore.go:116] using globspec upload/src/.s2iignore 
I0807 07:29:38.261885  100867 ignore.go:122] Glob result upload/src/.s2iignore 
I0807 07:29:38.261888  100867 ignore.go:84] .s2iignore lists a file spec of Dockerfile.s2i-fake 
I0807 07:29:38.261891  100867 ignore.go:116] using globspec upload/src/Dockerfile.s2i-fake 
I0807 07:29:38.261896  100867 ignore.go:84] .s2iignore lists a file spec of LICENSE 
I0807 07:29:38.261899  100867 ignore.go:116] using globspec upload/src/LICENSE 
I0807 07:29:38.261904  100867 ignore.go:84] .s2iignore lists a file spec of README.md 
I0807 07:29:38.261907  100867 ignore.go:116] using globspec upload/src/README.md 
I0807 07:29:38.261911  100867 ignore.go:122] Glob result upload/src/README.md 
I0807 07:29:38.261915  100867 ignore.go:84] .s2iignore lists a file spec of vendor 
I0807 07:29:38.261918  100867 ignore.go:116] using globspec upload/src/vendor 
I0807 07:29:38.261925  100867 ignore.go:45] attempting to remove file upload/src/.env.example 
I0807 07:29:38.261951  100867 ignore.go:45] attempting to remove file upload/src/.git 
I0807 07:29:38.264608  100867 ignore.go:45] attempting to remove file upload/src/.s2iignore 
I0807 07:29:38.264624  100867 ignore.go:45] attempting to remove file upload/src/README.md 
I0807 07:29:38.264636  100867 ignore.go:45] attempting to remove file upload/src/.dockerignore 
I0807 07:29:38.264648  100867 dockerfile.go:115] Constructing image build context directory at .
I0807 07:29:38.264663  100867 dockerfile.go:456] found override script file assemble
I0807 07:29:38.264667  100867 dockerfile.go:456] found override script file run
I0807 07:29:38.264689  100867 environment.go:41] Setting 7 environment variables provided by environment file in sources
I0807 07:29:38.264699  100867 dockerfile.go:209] Override scripts are included in directory "upload/scripts"
I0807 07:29:38.264705  100867 dockerfile.go:223] Processing injected inputs: api.VolumeList(nil)
I0807 07:29:38.264709  100867 dockerfile.go:225] Processed injected inputs: api.VolumeList(nil)
I0807 07:29:38.264845  100867 dockerfile.go:296] Wrote custom Dockerfile to Containerfile.valid
I0807 07:29:38.264850  100867 build.go:180] Application dockerfile generated in Containerfile.valid
```

`$ cat Containerfile.valid `

```dockerfile
FROM registry.access.redhat.com/ubi8/php-83:latest
LABEL "io.openshift.s2i.build.image"="registry.access.redhat.com/ubi8/php-83:latest" \
      "io.openshift.s2i.build.commit.author"="Stephen Cuppett <steve@cuppett.com>" \
      "io.openshift.s2i.build.commit.date"="Sun Mar 20 08:16:49 2022 -0400" \
      "io.openshift.s2i.build.commit.id"="fefad4475746a38a4197f36bff312dd056155f48" \
      "io.openshift.s2i.build.commit.ref"="master" \
      "io.openshift.s2i.build.commit.message"="Upgrade Laravel 9 (Laravel 8.68.1 -> 9.5.1)" \
      "io.openshift.s2i.build.source-location"="/home/scuppett/Development/php/laravel-s2i/."
ENV DOCUMENTROOT="/public" \
    APP_NAME="Laravel" \
    APP_DEBUG="true" \
    APP_URL="http://localhost:8080" \
    LOG_CHANNEL="errorlog" \
    DB_CONNECTION="sqlite" \
    DB_DATABASE=":memory:"
USER root
# Copying in override assemble/run scripts
COPY upload/scripts /tmp/scripts
# Copying in source code
COPY upload/src /tmp/src
# Change file ownership to the assemble user. Builder image must support chown command.
RUN chown -R 1001:0 /tmp/scripts /tmp/src
USER 1001
RUN /tmp/scripts/assemble
CMD /tmp/scripts/run
```

I'm not sure if I'm invoking this right or if I'm missing some other key attribute or behavior here.

### PR DESCRIPTION
**Pull request by cuppett**
Whether the base image provides the scripts or not, the repository can supply override scripts. Removes the expectation/need of a location specified in the image.

### CODE REVIEW

**Comment by bparees** [comment]
i think we're lacking some fundamental clarity about the precedence ordering of scripts.  

i think it should be:

1) use the scripts url the user provided in the config/command line (weirdly we seem to have two args for this `--image-script-url` and `--scripts-url` (as well as the deprecated `--scripts`).  I think the intent was that image-scripts-url was supposed to point to a location within the image while scripts-url could point pretty much anywhere, but i'm not seeing the code actually differentiate on those two args in that way.

2) use the scripts found in the source repo (this needs to be second to (1) so that a user can do a build in which they override whatever scripts the repo might happen to provide)

3) use the scripts url specified by the builder image in the builder image's labels.

The existing code, prior to this PR, does not prioritize scripts in the repository.  So in that sense this is a good change.

However the change also now prioritizes scripts from the repo over whatever config/args the user might have provided at runtime to indicate where they want to get their scripts from, meaning that you have to delete the scripts from your repository to be able to use an alternate location.  Which potentially changes the behavior for existing users who are passing `--scripts-url` because that arg will now have no effect. (use of `--image-scripts-url` is, i think, unaffected because getImageScripts was already returning "false" for that arg, meaning we'd still scan the repo for scripts and use them, which is at least one of the differences between --image-scripts-url and --scripts-url).

tldr: i think this needs a little more thought/study to straighten out the interactions of the various ways of providing scripts.  The other subtly here is that there's an intent to union together the sources of scripts (e.g. you might get your assemble script from the local repo, but the run script might still come from the builder image).  Looking at this code(both the original and the patch) i'm not sure how all that comes together at the moment.

we also need to compare this behavior with the behavior when we directly build the image (the code path being modified here is only for the "create a dockerfile" mode, not the "invoke docker build mode" which is somewhat different) and ensure we're being consistent in how scripts are chosen between the two modes.

**Comment by cuppett** [comment]
> 1. use the scripts url the user provided in the config/command line (weirdly we seem to have two args for this `--image-script-url` and `--scripts-url` (as well as the deprecated `--scripts`).  I think the intent was that image-scripts-url was supposed to point to a location within the image while scripts-url could point pretty much anywhere, but i'm not seeing the code actually differentiate on those two args in that way.

In my s2i command I was using to test this, I didn't specify any command line args like that (only had the annotation in the image and the stuff in the local repo). I didn't see where these would be distinguishable in the fragment of code I was looking at; it must be further upstream where that is lost.

I'll get back to this if somebody else doesn't first.

/hold if we can (I'm looking for the WIP toggle too)

**Comment by cuppett** [comment]
Have reworked this PR to honor the documented ordering and pick up the .s2i scripts in the local repository in the right spot. Re-ran my checks and saw the right behaviors.

I noticed that annotations or command line parameters which might use http:// or file:// aren't really considered/used in the dockerfile.go (they could be checked/fetched and merged into the resolved list), but I didn't try to resolve any of that.

**Comment by coreydaley** [comment]
/assign @bparees 
Do you want to give this a once-over since you had a very opinionated comment about how it should work?

**Comment by bparees** [comment]
> Do you want to give this a once-over since you had a very opinionated comment about how it should work?

i'd prefer you internalize the summary i gave of how this is supposed to work and assess whether this PR achieves that behavior (or ask me questions about my summary if you think it's incorrect or unclear in the expectations it sets).

File: pkg/build/strategies/dockerfile/dockerfile.go [hunk 1]
```diff
@@ -424,18 +420,27 @@ func getDestination(config *api.Config) string {
 
 // getImageScriptsDir returns the directory containing the builder image scripts and a bool
 // indicating that the directory is expected to contain all s2i scripts
-func getImageScriptsDir(config *api.Config) (string, bool) {
+func getImageScriptsDir(config *api.Config, builder *Dockerfile) (string, map[string]bool) {
```
Comment (coreydaley) [inline pkg/build/strategies/dockerfile/dockerfile.go:423]: The comment for this function does not accurately describe the function and needs to be updated.

File: pkg/build/strategies/dockerfile/dockerfile.go [hunk 2]
```diff
@@ -422,20 +418,29 @@ func getDestination(config *api.Config) string {
...
+
+	// 3rd priority (the builder image), collect the locations
 	scriptsURL, _ := util.AdjustConfigWithImageLabels(config)
 	if strings.HasPrefix(scriptsURL, "image://") {
-		return strings.TrimPrefix(scriptsURL, "image://"), true
+		return strings.TrimPrefix(scriptsURL, "image://"), providedScripts
 	}
 	if strings.HasPrefix(config.ImageScriptsURL, "image://") {
-		return strings.TrimPrefix(config.ImageScriptsURL, "image://"), false
+		return strings.TrimPrefix(config.ImageScriptsURL, "image://"), providedScripts
```
Comment (coreydaley) [inline pkg/build/strategies/dockerfile/dockerfile.go:439]: Shouldn't this return `make(map[string]bool)` instead according to it's previous use?

File: pkg/build/strategies/dockerfile/dockerfile_test.go [hunk 3]
```diff
@@ -80,27 +77,22 @@ func TestGetImageScriptsDir(t *testing.T) {
...
 				},
 			},
-			ExpectedDir:   "/usr/some/dir",
-			HasAllScripts: true,
+			ExpectedDir: "/usr/some/dir",
 		},
 	}
 	for _, tc := range cases {
-		output, hasScripts := getImageScriptsDir(tc.Config)
+		output, _ := getImageScriptsDir(tc.Config, &(Dockerfile{}))
```
Comment (coreydaley) [inline pkg/build/strategies/dockerfile/dockerfile_test.go:92]: Shouldn't some of these tests check for the returned map and it's contents (or lack thereof) since the `hasScripts` bool is being changed?

**Review by coreydaley — CHANGES_REQUESTED** [review]

File: pkg/build/strategies/dockerfile/dockerfile.go (hunk 2 above)
Comment (cuppett) [reply pkg/build/strategies/dockerfile/dockerfile.go:439]: I don't think so. This function has been expanded to merge the precedence order (which was the issue with the previous version). 

In this case, when scripts have been found in the local repo, but then in the 3rd and 4th priority cases where an annotation or fallback exists, we should merge/override with the scripts from the local repo.

The only correct spot to return an empty found map is when we have a command line parameter pointing to image:/// location (because we won't know and don't inspect the image to see *which* scripts may be there, so it's unilateral).

It'd be worthwhile in a future enhancement to understand more deeply how we should handle or inspect when file:/// or http:// are indicated for these ordered search locations. I couldn't see where that it's inspected anywhere and didn't try it.

File: pkg/build/strategies/dockerfile/dockerfile_test.go (hunk 3 above)
Comment (cuppett) [reply pkg/build/strategies/dockerfile/dockerfile_test.go:92]: The tests here were pretty deficient from what I saw. It only tested a "true" condition and empty map without providing any file-based overrides. 

I could revise this to plop &(Dockerfile{}) in the stanzas like the "true" previously; however, it would just confuse reviewers again next time that there is an actual validation here (versus it being static/empty in all cases). I think this representation more accurately matches the current and previous value of what the test was providing us.

I looked briefly at how to mock filesystem entries but I think I'll make a bigger mess before I provide more value.
